src/proxy-ui-api/src/main/resources/openapi-definition.yaml

```diff
@@ -1555,14 +1708,29 @@ paths:
           description: resource requested does not exists
         '406':
           description: request specified an invalid format
+        '409':
+          description: an existing item already exists
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ErrorInfo'
+              examples:
+                error_metadata_response:
+                  $ref: '#/components/examples/ErrorWithMetadataExample'
         '500':
           description: internal server error
     post: # ok
       tags:
         - clients
       summary: add new service description for the security server client
       operationId: addClientServiceDescription
-      description: Administrator adds a new service description for the client.
+      description: >
+        Administrator adds a new service description for the client. This endpoint can also return metadata in the error response.
+        The metadata array can contain error messages about why adding the service description did not succeed.
```

Comment:
The metadata for this and the Security Server initialization are now explained in the api doc.

Reply:
Metadata structure for adding service description and initialization is well explained in the openapi definition, nice!

However there are some endpoints where warning metadata structure is not explained. I think those should also be documented according to our decisions on public/private API documentation

we need to document warning and error metadata for external clients. Primary try to do this in openapi definition if feasible, if not, some other way

Warning metadata structure does not seem to be explained for
- upload backup
- add (unregistered) client

Not sure what the status is for _error_ metadata structures? 

Did we already have a blanket statement about metadata being "subject to change"?

> ...but with the disclaimer that these details are subject to change at any time

Reply:
Metadata structure for service_already_exists and initialization errors is well explained in the openapi definition, nice!

However there are some other endpoints where warning metadata structure is not explained. I think those should also be documented according to our decisions on public/private API documentation

we need to document warning and error metadata for external clients. Primary try to do this in openapi definition if feasible, if not, some other way

Warning metadata structure does not seem to be explained for

upload backup
add (unregistered) client
add service description warnings (validation and add/remove services)
initialization warnings that have metadata
delete key
update, refresh service description (copypaste from add, probably)
Some of these are pretty simple. Like delete key, additional sentence "...returns metadata that contains key id" should be enough probably.

Not sure what the status is for error metadata structures? Are all relevant ones documented?

Did we already have a blanket statement about metadata being "subject to change"?

...but with the disclaimer that these details are subject to change at any time

Reply:
IIRC the error codes were not originally intended to be published in the API definition but rather in a separate document? 

I think it might make sense to explain the metadata of e.g. the service description endpoints which might be a bit complex. At the moment the metadata is simply declared like this 
```
<p>This endpoint can also return metadata in the error response. The metadata array can contain error messages about why adding the service description did not succeed (for example invalid URL's or existing duplicate service codes).</p>
```
The error code itself was omitted on purpose and a more generic example was included.

Reply:
> Did we already have a blanket statement about metadata being "subject to change"?

We did not. Wasn't it supposed to be in a separate document with all possible error codes?

Sure we can add it e.g. to the info section of the API definition or perhaps to all endpoints separately?

Reply:
Updated the descriptions in the suggested places. Review with caution since it's becoming harder to keep track of all the changes.

Reply:
Should the disclaimer be added to the API document somewhere?

Reply:
I think there may be a bit of misunderstanding about error / warning metadata. Metadata shouldn't really be "error messages" as such, it should be raw data that adds more detail to the error, when that is needed. It may be used for generating error messages, but maybe also for other purposes.  


E.g. in the case of duplicate service

```
{
   "status": 409,
   "error": {
     "code": "service_already_exists",
     "metadata": [
       "duplicateService.v1",
       "http://xroad-lxd-cs.lxd/duplicate1.wsdl"
     ]
   }
 }
```

metadata contains service codes and service wsdls. Those can be used to display nice UI error messages, but they could also be used programmatically by some API client when he creates some maintenance scripts. For example 

IF (service_already_exists && !    
  wsdls_reported.contains(extract_wsdls_from_metadata())
    send_email_to_wsdl_admin(extract_wsdls_from_metadata()) 

To be able to use the metadata properly, it's structure needs to be documented. e.g. API client script programmer needs to know if metadata contains

```
     "metadata": [
       "fooService.v1",
       "http://xroad-lxd-cs.lxd/foo.wsdl",
       "barService.v1",
       "http://xroad-lxd-cs.lxd/bar.wsdl"
     ]
```

or

```
     "metadata": [
       "fooService.v1",
       "barService.v1",
       "http://xroad-lxd-cs.lxd/foo.wsdl",
       "http://xroad-lxd-cs.lxd/bar.wsdl"
     ]
```

Since it is so complex, and documenting is somewhat clumsy, metadata should be used sparingly. Metadata should only be used when there is a clear need, and same information is not available to client in other ways (e.g. request parameters that client has sent)

So I think we have decisions https://confluence.niis.org/display/XRDDEV/Public+and+private+API+contracts%2C+API+versioning%2C+error+codes%2C+warnings+and+metadata#PublicandprivateAPIcontracts,APIversioning,errorcodes,warningsandmetadata-Conclusion/decisions

- error codes - ErrorCodes.java
- error metadata - openapi-definition.yaml
- warning codes - no clear decision I guess, openapi-definition.yaml, openapi-definition.yaml examples, ErrorCodes.java, WarningCodes.java....?
- warning metadata - openapi-definition.yaml

Reply:
> Should the disclaimer be added to the API document somewhere?


I think it would be good to have. You had good ideas about how to do it in an earlier comment.

Reply:
I looked at the warning metadata listed earlier (upload backup...update, refresh service description) - documentation looks spot on :+1:  so I think these are taken care of:

- warning codes - openapi-definition.yaml
- warning metadata - openapi-definition.yaml

Reply:
Updated error metadatas and added a note to the info stating that the metadatas might be changed